### ISSUE TITLE
Why is classdef import treated differently?

### ISSUE THREAD

**Issue by MrMino** [ericdill/depfinder#84]
Hi there 👋.

I'm looking into `depfinder` as a possible substitute for `imp` module that we currently use in `pip-check-reqs`.

There's one thing that I'm unsure about, namely the fact that `depfinder` treats class definitions differently from global code:

```py
class X:
    import x.y.z
```

```
$ depfinder module.py
{'questionable': ['x']}
```
Is `questionable` meant to denote imports that may or may not be executed?
If so, why is this import being marked as such? It's going to be executed, whether `X` is instantiated or not.

**Comment by ocefpaf** [comment]
@ericdill can probably answer this better but in my view this is b/c a hidden import may mean an optional dependency. Depfinder was designed to help user create requirements.txt files for library authors and packagers and, in those cases, optional dependencies need to be checked and treated differently then mandatory ones.

**Comment by MrMino** [comment]
I understand the import optionality idea :). 

The issue is that, while imports inside `try..except` or `def` might never get executed, `class` has no bearing on whether something gets executed or not; if execution gets to the line with `class`, all of the stuff inside it will get executed, always.

E.g.:

```
class Hello:
    print("Hello world")
```

Will execute the print as soon as the class gets defined. An import inside a class definition would too.

**Comment by ericdill** [comment]
Sounds like we should add this use case to the “required” key

**Comment by ocefpaf** [comment]
> Looks like I didn’t fully understand what gets executed and when.

I guess that this is not a clear definition b/c folks use these patterns very loosely. For example, I always saw import inside classes and functions as "hidden" and hence optional/activated only when that function/class is used. Mandatory ones are at the top. However, many use that pattern to delay the import and make root imports faster.

TL;DR I don't think you didn't understand. This touches the topic on how folks write code and makes it really hard to define :-)

**Comment by ericdill** [comment]
My original intent with this library is a coarse-grained separation of “things that are required to import the module” with “everything else that might be required at runtime”. Those two categories are now called “required” and “questionable”, respectively. It seems that the example brought by @MrMino falls within the “requires to import the module” category for sure.  

> I don't think you didn't understand

/shrug. I have no problem being publicly wrong about something 🥳. It’s an opportunity to learn something new

**Comment by MrMino** [comment]
> Sounds like we should add this use case to the “required” key. Shouldn’t be too hard to implement

@ericdill  I suppose it just requires taking out `ast.ClassDef` from `AST_QUESTIONABLE`.

>  This touches the topic on how folks write code and makes it really hard to define

@ocefpaf Well... sure, but if the intent of the user is to make an import optional, then placing it into class definition alone doesn't do that. In fact, in that case, `depfinder` might be misleading users into thinking that it does.

The only two things that `class` does is it changes the way `def` behaves and constructs a new namespace. With regards to import statements however, it behaves in no way different than `with` does. The import doesn't get propagated outside of the namespace created by `class`, but there's no conditionality.

**Comment by ericdill** [comment]
> [@ericdill](https://github.com/ericdill) I suppose it just requires taking out ast.ClassDef from AST_QUESTIONABLE

Yea. As long as it’s only ast.ClassDef I think we are okay to move that into the REQUIRED. I assume you can do horrible things like conditionally import within an if/else block inside a ClassDef? If so, we would want to get that marked as questionable. And we still want to mark imports inside a class function as questionable. Any other conditions that are worth testing for explicitly?

**Comment by MrMino** [comment]
Nothing pops up to my mind at this time, apart from the usual suspects: `try..except`, `if..elif..else`, `def`.

You could treat the code inside the `ClassDef` the same way as statements in the global scope of a module. AFAIK that's what they are to the Python compiler.

**Comment by MrMino** [comment]
FWIW I think you're also missing `ast.AsyncFunctionDef`.

**Comment by MrMino** [comment]
Aaand if you are going to be formal about it, then also `ast.For` and `ast.AsyncFor`, since the body of a `for` loop does not execute if the iterator is empty.

:smile:

**Comment by ericdill** [comment]
Nice. Those are all great. You have any interest in opening a PR?

**Comment by MrMino** [comment]
I'm not sure if I know the code base well enough, but I can get something started.

**Comment by ericdill** [comment]
> I'm not sure if I know the code base well enough, but I can get something started.

ok awesome! let me know how I can help. more eyes makes a better project after all
